x/bank/types/keys.go

```diff
@@ -22,58 +22,38 @@ const (
 var (
 	SupplyKey           = collections.NewPrefix(0)
 	DenomMetadataPrefix = collections.NewPrefix(1)
-	DenomAddressPrefix  = []byte{0x03}
-
 	// BalancesPrefix is the prefix for the account balances store. We use a byte
 	// (instead of `[]byte("balances")` to save some disk space).
-	BalancesPrefix = []byte{0x02}
-
+	BalancesPrefix     = collections.NewPrefix(2)
+	DenomAddressPrefix = collections.NewPrefix(3)
 	// SendEnabledPrefix is the prefix for the SendDisabled flags for a Denom.
 	SendEnabledPrefix = collections.NewPrefix(4)
 
 	// ParamsKey is the prefix for x/bank parameters
 	ParamsKey = collections.NewPrefix(5)
 )
 
-// AddressAndDenomFromBalancesStore returns an account address and denom from a balances prefix
-// store. The key must not contain the prefix BalancesPrefix as the prefix store
-// iterator discards the actual prefix.
-//
-// If invalid key is passed, AddressAndDenomFromBalancesStore returns ErrInvalidKey.
-func AddressAndDenomFromBalancesStore(key []byte) (sdk.AccAddress, string, error) {
-	if len(key) == 0 {
-		return nil, "", ErrInvalidKey
-	}
-
-	kv.AssertKeyAtLeastLength(key, 1)
-
-	addrBound := int(key[0])
-
-	if len(key)-1 < addrBound {
-		return nil, "", ErrInvalidKey
+// NewBalanceCompatValueCodec is a codec for encoding Balances in a backwards compatible way
+// with respect to the old format.
+func NewBalanceCompatValueCodec() collcodec.ValueCodec[math.Int] {
```

Comment:
We have created a special bank balance value codec, which in the end it's a `math.Int`  value but can also return `math.Int` in case the bytes are a `sdk.Coin`, this retains backwards compatibility.